### ISSUE TITLE
`Literal` `str`s do not work with `Sequence[T]` generic function argument narrowing

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#1628]
### Summary

Minimization of when I ran into this in my code:

https://play.ty.dev/21a7ef1b-756b-4981-af2d-878f5fe72ee4
```py
from typing import Sequence, assert_type

def first[T](x: Sequence[T]) -> T:
    return x[0]

reveal_type("1")  # Revealed type: `Literal["1"]`
assert_type(first("1"), str)  # Type `str` does not match asserted type `Unknown`
```
Since the type of `"1"` is defaulted to `Literal["1"]`, code that tries to rely on literal strings with this `Sequence` generic narrowing will not work.

### Version

ty 0.0.1-alpha.27 (26d7b6864 2025-11-18)

**Comment by carljm** [comment]
Yes, this is odd, since `str` directly inherits `Sequence[str]`. It looks like we aren't falling back from `Literal["1"]` to `str` correctly?

**Comment by MeGaGiGaGon** [comment]
I think it has something to do with the generic, since `static_assert(is_assignable_to(Literal["1"], Sequence[str]))` works as expected, so the generic is somehow causing the fallback to not apply/be tried/fail.

**Comment by carljm** [comment]
Yes, we definitely understand the relationship between `Literal["1"]` and `str` in general, I mean that in the typevar solver we aren't falling back where we need to.

**Comment by nklsla** [comment]
I think this one might be related to my issue.
I get an error, `ty[invalid-argument-type]`, from my function signature when using typehint. Removing the typehint resolves the error. Also defining the variable inside the function with or without the typehint results in no error either.
Im using `0.0.1a29`

```py
import polars as pl

def winsorize(
    col: pl.Expr | str,
    lower_quantile: float = 0.05,
    upper_quantile: float = 0.95,
    method: str = "linear", # Using typehint results in this error
    # method = "linear",  # No error, however not using typehint
) -> pl.Expr:
    """Winsorize a column by clipping its values to the specified quantiles."""

    if isinstance(col, str):
        _col = pl.col(col)
    elif isinstance(col, pl.Expr):
        _col = col
    else:
        raise TypeError("col must be a str or pl.Expr")

#  Re-/defining the value in function gives not error
#  method = "linear"  # no error
#  method: str = "linear" # no error
    return _col.clip(
        lower_bound=_col.quantile(lower_quantile, interpolation=method), # Argument to bound method `quantile` is incorrect: Expected `Literal["nearest", "higher", "lower", "midpoint", "linear", "equiprobable"]`, found `str`ty[invalid-argument-type](https://ty.dev/rules#invalid-argument-type)
        upper_bound=_col.quantile(upper_quantile, interpolation=method), # Argument to bound method `quantile` is incorrect: Expected `Literal["nearest", "higher", "lower", "midpoint", "linear", "equiprobable"]`, found `str`ty[invalid-argument-type](https://ty.dev/rules#invalid-argument-type)
    )
```

This is the signature for the `polars column` method `quantile`
```py
bound method Expr.quantile(
    quantile: int | float | Expr,
    interpolation: Literal["nearest", "higher", "lower", "midpoint", "linear", "equiprobable"] = Literal["nearest"]
) -> Expr
```

**Comment by sharkdp** [comment]
> I get an error, `ty[invalid-argument-type]`, from my function signature when using typehint.

I don't think this is related to the issue here. If you annotate `method` as `str`, then ty *should* error on your code, because `Expr.quantile` expects `Literal["nearest", "higher", "lower", "midpoint", "linear", "equiprobable"]`. `str` is a supertype of this union of literals, and so you might accidentally pass invalid `str` values to `Expr.quantile`. You probably want to annotate your `method` parameter in the same way that `Expr.quantile`'s `interpolation` parameter is annotated.

**Comment by MeGaGiGaGon** [comment]
Looks like this was fixed in the 0.0.18 release:
```
PS ~>echo @'
from typing import Sequence, assert_type, reveal_type

def first[T](x: Sequence[T]) -> T:
    return x[0]

reveal_type(first("1"))
'@ > issue.py
PS ~>uvx ty@0.0.17 check issue.py
info[revealed-type]: Revealed type
 --> issue.py:6:13
  |
4 |     return x[0]
5 |
6 | reveal_type(first("1"))
  |             ^^^^^^^^^^ `Unknown`
  |

Found 1 diagnostic
PS ~>uvx ty@0.0.18 check issue.py
info[revealed-type]: Revealed type
 --> issue.py:6:13
  |
4 |     return x[0]
5 |
6 | reveal_type(first("1"))
  |             ^^^^^^^^^^ `str`
  |

Found 1 diagnostic
```
